### PR TITLE
Review

### PR DESCRIPTION
**Pull request by VictoryKS**

### CODE REVIEW

File: .eslintrc.yml [hunk 1]
```diff
@@ -0,0 +1,173 @@
+env:
+  es6: true
+  node: true
+extends: 'eslint:recommended'
+globals:
+  api: false
```
Comment (tshemsedinov) [inline .eslintrc.yml]: Remove global api variable

File: JavaScript/queuing-system.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+function Queue(userName) {
+  this.userName = userName;
+  this.first = null;
+  this.last = null;
+  this.length = 0;
+  this.time = 0;
+}
+
+Queue.prototype.timeout = function(time) {
```
Comment (tshemsedinov) [inline JavaScript/queuing-system.js]: `setTimrout(msec)`

File: JavaScript/queuing-system.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
+'use strict';
+
+function Queue(userName) {
+  this.userName = userName;
+  this.first = null;
+  this.last = null;
+  this.length = 0;
+  this.time = 0;
```
Comment (tshemsedinov) [inline JavaScript/queuing-system.js]: `this.timeout`

File: JavaScript/queuing-system.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    }
+  }
+  this.length++;
+  if (this.time) {
+    setTimeout(() => {
+      if (this.length <= 1) {
+        this.first = null;
+        this.last = null;
+        this.length = 0;
+      } else {
```
Comment (tshemsedinov) [inline JavaScript/queuing-system.js]: `> 1`

File: JavaScript/queuing-system.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+};
+
+function QueuingSystem() {
+  this.queues = [];
+}
+
+QueuingSystem.prototype.logIn = function(userName) {
+  let contains = false;
+  this.queues.forEach(q => {
+    if (q.userName === userName) contains = true;
```
Comment (tshemsedinov) [inline JavaScript/queuing-system.js]: `return`

File: JavaScript/queuing-system.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    this.queues.push(list);
+    return list;
+  }
+};
+
+QueuingSystem.prototype.logOff = function(userName) {
+  const q = this.queues;
+  for (let i = 0; i < q.length; i++) {
+    if (q[i].userName === userName) {
+      q.splice(i, 1);
```
Comment (tshemsedinov) [inline JavaScript/queuing-system.js]: `return`

**Review by tshemsedinov — CHANGES_REQUESTED** [review]
:+1:

File: JavaScript/queuing-system.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  this.timeout = 0;
+}
+
+Queue.prototype.setTimeout = function(msec) {
+  this.timeout = msec;
+  return this;
+};
+
+Queue.prototype.add = function(item, priority) {
+  let element = { prev: null, next: null, item, priority };
```
Comment (tshemsedinov) [inline JavaScript/queuing-system.js]: Better use `const` here

File: JavaScript/queuing-system.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  this.length = 0;
+  this.timeout = 0;
+}
+
+Queue.prototype.setTimeout = function(msec) {
+  this.timeout = msec;
+  return this;
+};
+
+Queue.prototype.add = function(item, priority) {
```
Comment (tshemsedinov) [inline JavaScript/queuing-system.js]: It would be great to decompose this function into 2-3 smaller

File: JavaScript/queuing-system.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+      element.prev = current;
+      if (!element.next) this.last = element;
+    }
+  }
+  this.length++;
+  if (this.timeout) {
+    setTimeout(() => {
+      if (this.length > 1) {
+        this.last = element.prev;
+        element = null;
```
Comment (tshemsedinov) [inline JavaScript/queuing-system.js]: Why?

File: JavaScript/queuing-system.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+  }
+  this.length--;
+  return element.item;
+};
+
+Queue.prototype.drain = function() {
+  this.first = null;
+  this.last = null;
+  this.length = 0;
+  this.time = 0;
```
Comment (tshemsedinov) [inline JavaScript/queuing-system.js]: Do you remember that we renamed `time` to `timeout`?

File: JavaScript/queuing-system.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+};
+
+function QueuingSystem() {
+  this.queues = [];
+}
+
+QueuingSystem.prototype.logIn = function(userName) {
+  this.queues.forEach(q => {
```
Comment (tshemsedinov) [inline JavaScript/queuing-system.js]: I'd prefer for here

File: JavaScript/queuing-system.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    if (q.userName === userName) return null;
+  });
+  const list = new Queue(userName);
+  this.queues.push(list);
+  return list;
+};
+
+QueuingSystem.prototype.logOff = function(userName) {
+  const q = this.queues;
```
Comment (tshemsedinov) [inline JavaScript/queuing-system.js]: `const queues = this.queues;`

File: JavaScript/usage.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+const User1 = q.logIn('John');
+const User2 = q.logIn('Tom');
+const User3 = q.logIn('John');
+
+const obj1 = { number: 'first' };
+const obj2 = { number: 'second' };
+const obj3 = { number: 'third' };
+
+User1.setTimeout(3000)
+     .add(obj1, 7)
```
Comment (tshemsedinov) [inline JavaScript/usage.js]: 2 space indentation

File: JavaScript/usage.js [hunk 14]
```diff
@@ -0,0 +1,39 @@
+'use strict';
+
+const QueuingSystem = require('./queuing-system.js');
+
+const q = new QueuingSystem();
```
Comment (tshemsedinov) [inline JavaScript/usage.js]: Better naming `qsys`

File: JavaScript/usage.js [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+const User1 = q.logIn('John');
```
Comment (tshemsedinov) [inline JavaScript/usage.js]: Naming: instances should be named in `lowerCamelCase`, so `user1`...

**Review by tshemsedinov — CHANGES_REQUESTED** [review]
:+1: :+1:

File: JavaScript/queuing-system.js (hunk 9 above)
Comment (VictoryKS) [reply JavaScript/queuing-system.js]: After some time I delete the last element, if it is still in the queue

Comment (VictoryKS) [reply JavaScript/queuing-system.js]: Oh, it's my fault. I guess it should be:
`this.last = this.last.prev;`
`this.last.next = null;`

**Review by tshemsedinov — APPROVED** [review]

**Pull request merged**